crates/bevy_ecs/macros/src/fetch.rs

```diff
@@ -379,64 +358,89 @@ pub fn derive_world_query_impl(ast: DeriveInput) -> TokenStream {
         }
     };
 
-    let read_only_impl = if fetch_struct_attributes.is_filter {
-        quote! {}
-    } else if fetch_struct_attributes.is_mutable {
-        let fetch_impl = impl_fetch(
+    let read_only_fetch_impl = match fetch_struct_attributes.is_mutable {
+        true => impl_fetch(
             false,
-            read_only_fetch_associated_type,
+            true,
             read_only_fetch_struct_name.clone(),
-            read_only_item_struct_name.clone(),
-        );
-
-        quote! {
-            #fetch_impl
+            read_only_item_struct_name,
+        ),
+        false => quote! {},
+    };
 
-            impl #impl_generics #path::query::WorldQuery for #read_only_item_struct_name #ty_generics #where_clause {
-                type Fetch = #read_only_fetch_struct_name #ty_generics;
-                type State = #state_struct_name #ty_generics;
-                type ReadOnlyFetch = #read_only_fetch_struct_name #ty_generics;
+    let fn_shrink_internals = match fetch_struct_attributes.is_filter {
+        true => quote! { item },
+        false => quote! {
+            #item_struct_name {
+                #(
+                   #field_idents : < #field_types as #path::query::WorldQuery> :: shrink( item.#field_idents ),
+                )*
+                #(
+                    #ignored_field_idents: item.#ignored_field_idents,
+                )*
             }
-        }
-    } else {
-        quote! {
-            // Statically checks that the safety guarantee actually holds true. We need this to make
-            // sure that we don't compile `ReadOnlyFetch` if our struct contains nested `WorldQuery`
-            // members that don't implement it.
-            #[allow(dead_code)]
-            const _: () = {
-                fn assert_readonly<T: #path::query::ReadOnlyFetch>() {}
-
-                // We generate a readonly assertion for every struct member.
-                fn assert_all #impl_generics () #where_clause {
-                    #(assert_readonly::<<#field_types as #path::query::WorldQuery>::Fetch>();)*
-                }
-            };
-        }
+        },
+    };
+
+    let read_only_asserts = match fetch_struct_attributes.is_mutable {
+        true => quote! {
+            // this is technically unnecessary as `<_ as WorldQueryGats<'world>>::ReadOnlyFetch: ReadOnlyFetch<'world>`
+            // but to protect against future mistakes we assert the assoc type implements `ReadOnlyFetch` anyway
+            #( assert_readonly::<'__w, #path::query::ROQueryFetch<'__w, #field_types>>(); )*
+        },
+        false => quote! {
+            #( assert_readonly::<'__w, #path::query::QueryFetch<'__w, #field_types>>(); )*
```

Comment:
```suggestion
           // If it is mutable, we must check the `QueryFetch` type directly
            #( assert_readonly::<'__w, #path::query::QueryFetch<'__w, #field_types>>(); )*
```

Reply:
This feels... wrong, FWIW. Why is `fetch_struct_attributes.is_mutable == true` when the actual requested fetch is immutable? Why do we need to branch here, rather than always checking `QueryFetch`?

Reply:
@BoxyUwU probably knows more about this particular bit.